storage/src/main/java/com/tencent/rss/storage/handler/impl/HdfsClientReadHandler.java

```diff
@@ -51,7 +66,91 @@ public HdfsClientReadHandler(
         ShuffleStorageUtils.getShuffleDataPathWithRange(appId,
             shuffleId, partitionId, partitionNumPerRange, partitionNum));
     init(fullShufflePath);
-    readAllIndexSegments();
+    cursor = indexReaderMap
+        .entrySet()
+        .stream()
+        .sorted(Comparator.comparing(Entry::getKey))
+        .collect(Collectors.toList())
+        .iterator();
+  }
+
+  public ShuffleIndexResult readShuffleIndex() {
+    if (!cursor.hasNext()) {
+      return null;
+    }
+
+    Entry<String, HdfsFileReader> entry = cursor.next();
+    curPath = entry.getKey();
+    long start = System.currentTimeMillis();
+    HdfsFileReader reader = entry.getValue();
+    if (reader == null) {
+      LOG.error("Reader of {} is invalid!", curPath);
+      return null;
+    }
+
+    try {
+      byte[] indexData = reader.readAllData();
+      LOG.info("Read index files for shuffleId[{}], partitionId[{}] for {} ms",
+          shuffleId, partitionId, System.currentTimeMillis() - start);
+      return new ShuffleIndexResult(indexData);
+    } catch (Exception e) {
+      LOG.info("Fail to read index files for shuffleId[{}], partitionId[{}]", shuffleId, partitionId);
+    }
+
+    return new ShuffleIndexResult();
+  }
+
+  public ShuffleDataResult readShuffleData(ShuffleDataSegment shuffleDataSegment) {
+    HdfsFileReader reader = dataReaderMap.get(curPath);
+    if (reader == null) {
+      LOG.error("Invalid segment {}", shuffleDataSegment);
+      return null;
+    }
+
+    int expectedLength = shuffleDataSegment.getLength();
+    if (expectedLength <= 0) {
```

Comment:
if (shuffleDataSegment.getLength() < 0)

Reply:
OK

Reply:
expectedLength is useful in the following code